[compiler-rt] Add baremetal version of profile library.

efriedma-quic opened the pull request:
Adds a flag COMPILER_RT_PROFILE_BAREMETAL, which disables the parts of the profile runtime which require a filesystem or malloc.  This minimal library only requires string.h from the C library.

This is useful for profiling or code coverage of baremetal images, which don't have filesystem APIs, and might not have malloc configured (or have limited heap space).

Expected usage:

- Add code to your project to call `__llvm_profile_get_size_for_buffer()` and `__llvm_profile_write_buffer()` to write the profile data to a buffer in memory, and then copy that data off the device using target-specific tools.
- If you're using a linker script, set up your linker script to map the profiling and coverage input sections to corresponding output sections with the same name, and mark them KEEP.  `__llvm_covfun` and `__llvm_covmap` are non-allocatable, `__llvm_prf_names` is read-only allocatable, and `__llvm_prf_cnts` and `__llvm_prf_data` are read-write allocatable.
- The resulting data is in same format as the non-baremetal profiles.

There's some room for improvement here in the future for doing profiling and code coverage for baremetal. If we revised the profiling format, and introduced some additional host tooling, we could move some of the metadata into non-allocated sections, and construct the profraw file on the host.  But this patch is sufficient for some use-cases.

efriedma-quic committed:
[compiler-rt] Add baremetal version of profile library.

Adds a flag COMPILER_RT_PROFILE_BAREMETAL, which disables the parts of
the profile runtime which require a filesystem or malloc.  This minimal
library only requires string.h from the C library.

This is useful for profiling or code coverage of baremetal images, which
don't have filesystem APIs, and might not have malloc configured (or
have limited heap space).

There's some room for improvement here in the future for doing profiling
and code coverage for baremetal. If we revised the profiling format, and
introduced some additional host tooling, we could move some of the
metadata into non-allocated sections, and construct the profraw file on
the host.  But this patch is sufficient for some use-cases.

lenary commented:
It seems a little strange to me that baremetal profiling is using the "Linux" profiling platform (rather than "Other" or a new "Baremetal" platform). That could do with documenting a little bit better, I think.

I am not familiar enough with the profiling runtime to understand which bits you're disabling, but the individual changes seem well-reasoned?

smithp35 commented on compiler-rt/lib/profile/InstrProfilingMerge.c:
```diff
@@ -131,9 +130,11 @@ COMPILER_RT_VISIBILITY
 int __llvm_profile_merge_from_buffer(const char *ProfileData,
                                      uint64_t ProfileSize) {
   if (__llvm_profile_get_version() & VARIANT_MASK_TEMPORAL_PROF) {
+#ifndef COMPILER_RT_PROFILE_BAREMETAL
```
Could this be done in InstrProfilingPort.h ? By making the definitions of `PROF_ERR`, `PROF_WARN` and `PROF_NOTE` do nothing?

Probably more code, but maybe more maintainable.

efriedma-quic replied:
Done.

efriedma-quic commented:
> > There's some room for improvement here in the future for doing profiling and code coverage for baremetal. If we revised the profiling format, and introduced some additional host tooling, we could move some of the metadata into non-allocated sections, and construct the profraw file on the host. But this patch is sufficient for some use-cases.
> 
> How is this different from `-profile-correlate=binary`?

I wasn't aware of that.  I'll need to experiment.

That said, even if you strip out those sections, you still need to allocate a buffer large enough to hold a copy of all the counters, and run a bunch of code to fill that buffer.  Ideally you could just stop the CPU, use the debugger to copy some data into a file, and use that as input to llvm-prof.

------

> I'm fine with this change but I think a better strategy would be to start off with a minimal, header-only runtime implemented in C++ for baremetal systems which could also serve as a building block for full-featured OSes, eventually replacing the existing runtime which is rapidly approaching the point of being unmaintainable due to poor abstractions and code littered with #ifdefs.

I can see why you'd want a callback-driven interface to write out the profile file, instead of allocating a memory buffer or directly using filesystem APIs. But I don't see how any further granularity helps, unless you just want to avoid the compiler-rt build system.

------

> AND what about embedded systems based on musl libc (e.g. so called generic-none targets) if they don't define COMPILER_RT_PROFILE_BAREMETAL ? or do they now need to define COMPILER_RT_PROFILE_BAREMETAL (meaning many things)?

It maybe depends on what exactly you've done.  As far as I can tell, we currently use mmap, so I'm not sure how a baremetal target could build, unless your hacked up musl accidentally exposes mmap.  But if you have a setup that works, this patch shouldn't affect it.

It's possible that we should be checking for `__ELF__` in InstrProfilingPlatformLinux.c instead of `__linux__`; other places check for whether the target is ELF, instead of checking specific operating systems (including the LLVM instrumentation pass, and other places in libprofile).

efriedma-quic commented:
> If I've understood the scope of the change. The user/toolchain would still have to arrange for a mechanism to construct a means to output the counters in a format that llvm-profdata understands.

This subset of the profile library includes `__llvm_profile_write_buffer`, which outputs counters in the format llvm-profdata expects.  Obviously the user still needs to copy that data off the device somehow.

efriedma-quic committed:
Add comments explaining the different profile files.

Also adjust ifdef to be a bit more clear.

efriedma-quic commented:
Updated with some comments, and cleaned up one of the ifdefs a bit.

My thinking for the filenames is that we actually should split InstrProfilingPlatformLinux.c: we should have InstrProfilingPlatformELF.c for the ELF-only bits, and InstrProfilingPlatformStaticInit.c for the "generic" __start/__stop bits, and InstrProfilingPlatformWasm.c for the wasm-only bits (basically just a stub, but it doesn't fit anywhere else).  Then rename InstrProfilingPlatformOther.c to InstrProfilingPlatformDynamicInit.c.  But I'd prefer to leave that for a followup.

lenary commented:
Thanks for the additional comments.

I agree that splitting up these files better can wait for a follow-up commit.

I am happy, but don't feel familiar enough to approve this commit.

smithp35 commented:
Thanks for adding the comments. I agree with you on splitting the file, was tempted to mention that yesterday but it may have been a change too far for others to accept.

> This subset of the profile library includes __llvm_profile_write_buffer, which outputs counters in the format llvm-profdata expects. Obviously the user still needs to copy that data off the device somehow.

Thanks for pointing that out, it is quite easy to miss what the intended usage is. I eWhat I was thinking for a comment for toolchain vendors for how to use the API to save some time figuring it out. May not need to be in the source code, could put it in the description. 

Could it be as simple as:
```
extern uint64_t __llvm_profile_get_size_for_buffer(void);
extern int __llvm_profile_write_buffer(char *Buffer);

uint64_t size = __llvm_profile_get_size_for_buffer(void);
char* buffer = allocate_memory(size);
if (__llvm_profile_write_buffer(buffer) == -1) {
  // error
}
// Program specifc way to write out buffer
```

I'm with Lenary in that I'm supportive of the direction, but I don't have any prior commit experience. If we end up with no-one willing to approve/reject I can put my name to it though.

efriedma-quic committed:
Make sure profile lib builds if CMAKE_SYSTEM_NAME is Generic.

smithp35 reviewed (approved):
As discussed in the LLVM toolchains working group meeting. This looks like a worthwhile improvement for getting profiling working on bare-metal systems.

We agreed that any splitting of files can come in a follow up patch.

I have no further comments, but will be worth waiting a bit to see if any of the other reviewers have any.

efriedma-quic committed:
Define away PROF_ERR etc.

Simplifies uses a bit.

reactive-firewall reviewed (commented):
LGTM; with the already discussed caveat that there is room for future improvements. 

Specifically I like the additional comments for developers.

Hope this helps

efriedma-quic committed:
Fix formatting in compiler-rt/lib/profile/CMakeLists.txt

Co-authored-by: Petr Hosek <phosek@google.com>
